core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/aggregation/aggregators/Aggregators.kt

```diff
@@ -1,52 +1,136 @@
+        aggregate: Aggregate<Value, Return>,
+    ) = FlatteningAggregator.Factory(
+        getReturnTypeOrNull = getReturnTypeOrNull,
+        aggregator = aggregate,
+        preservesType = false,
+    )
+
+    /**
+     * Factory for a two-step aggregator that works only with numbers.
+     *
+     * @include [TwoStepNumbersAggregator]
+     */
+    private fun <Return : Number> twoStepForNumbers(
+        getReturnTypeOrNull: CalculateReturnTypeOrNull,
+        aggregate: Aggregate<Number, Return>,
+    ) = TwoStepNumbersAggregator.Factory(
+        getReturnTypeOrNull = getReturnTypeOrNull,
+        aggregate = aggregate,
+    )
+
+    /** @include [AggregatorOptionSwitch1] */
+    private fun <Param1, AggregatorType : Aggregator<*, *>> withOneOption(
+        getAggregator: (Param1) -> AggregatorProvider<AggregatorType>,
+    ) = AggregatorOptionSwitch1.Factory(getAggregator)
+
+    /** @include [AggregatorOptionSwitch2] */
+    private fun <Param1, Param2, AggregatorType : Aggregator<*, *>> withTwoOptions(
+        getAggregator: (Param1, Param2) -> AggregatorProvider<AggregatorType>,
+    ) = AggregatorOptionSwitch2.Factory(getAggregator)
+
+    // T: Comparable<T> -> T?
+    val min by twoStepPreservingType<Comparable<Any?>> {
+        minOrNull()
+    }
 
-    val max by preservesType<Comparable<Any?>> { maxOrNull() }
+    // T: Comparable<T> -> T?
+    val max by twoStepPreservingType<Comparable<Any?>> {
+        maxOrNull()
+    }
 
-    val std by withOption2<Boolean, Int, Number, Double> { skipNA, ddof ->
-        mergedValuesChangingTypes { std(it, skipNA, ddof) }
+    // T: Number? -> Double
+    val std by withTwoOptions { skipNA: Boolean, ddof: Int ->
+        flatteningChangingTypes<Number, Double>(stdTypeConversion) { type ->
+            std(type, skipNA, ddof)
+        }
     }
 
-    val mean by withOption<Boolean, Number, Double> { skipNA ->
-        changesType({ mean(it, skipNA) }) { mean(skipNA) }
+    // step one: T: Number? -> Double
+    // step two: Double -> Double
+    val mean by withOneOption { skipNA: Boolean ->
+        twoStepChangingType(
+            getReturnTypeOrNull = meanTypeConversion,
+            stepOneAggregator = { type -> mean(type, skipNA) },
+            stepTwoAggregator = { mean(skipNA) },
+        )
     }
 
-    val percentile by withOption<Double, Comparable<Any?>, Comparable<Any?>> { percentile ->
-        mergedValuesChangingTypes { type -> percentile(percentile, type) }
+    // T: Comparable<T>? -> T
+    val percentile by withOneOption { percentile: Double ->
+        flatteningPreservingTypes<Comparable<Any?>> { type ->
+            percentile(percentile, type)
+        }
     }
 
-    val median by mergedValues<Comparable<Any?>, Comparable<Any?>> { median(it) }
+    // T: Comparable<T>? -> T
+    val median by flatteningPreservingTypes<Comparable<Any?>> { type ->
+        median(type)
+    }
 
-    val sum by extendsNumbers { sum(it) }
+    // T: Number -> T
```

Comment:
// -> /** */ ?

Reply:
I will keep them like `//` here because it's just an implementation hint, not relevant from other places.